src/mpid/ch4/src/ch4_init.c

```diff
@@ -571,6 +571,10 @@ int MPID_Init(int requested, int *provided)
         MPIR_Assert(MPIDI_global.n_vcis <= MPIDI_CH4_MAX_VCIS);
         MPIR_Assert(MPIDI_global.n_vcis <= MPIR_REQUEST_NUM_POOLS);
     }
+#ifdef MPIDI_CH4_USE_WORK_QUEUES
+    /* workq does not work with multiple vci */
+    MPIR_Assert(MPIDI_global.n_vcis == 1);
+#endif
```

Comment:
I'm little unclear about this statement from the commit message "The only sensible benefit of workq is in the single vci case." You are not saying that the benefit of workq is limited to a single vci and we don't need support multi-vci later, are you?

Reply:
My reasoning is -- we  are enqueuing to workq handoff to reduce the overhead of issuing operations, so that we can reduce the thread contentions. With multiple VCI, the enqueuing  to workq itself will pose as a global serialization point (even with atomic enqueuing), thus voids all the benefit of multi-vci.

Reply:
The implementation can have one workq per vci. That's what we have in our implementation. That way we can operate the different vcis independent of each other.

Reply:
I see. I'll do the same then.

Reply:
@sagarth Currently it is difficult to spawn per-vci workq. As in current "main" branch, the workq lives ch4-layer, but vci is determined by lower layers.  How does it work if workq decide certain vci but later netmod decide to fall-back to active message?

Anyway, I think I would like to postpone multiple vci workq separately from this PR.

Reply:
Just for the record: we discussed about it offline.